### ISSUE TITLE
InMemory provider updates when reporting failed save

### ISSUE THREAD

**Issue by amosonn** [dotnet/efcore#24235]
## Description
When trying to save a data model with a duplicate Id, InMemory provider correctly throws an exception, but then still updates the state - unless data model is already in cache.

### `MyDbContext.cs`

```C#
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;

namespace test_efcore_inmemory
{
    /// <summary>
    /// Represents a session with the database.
    /// </summary>
    public class MyDbContext : DbContext
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="MyDbContext"/> class.
        /// </summary>
        /// <param name="options">The options to be used by the underlying DbContext.</param>
        public MyDbContext(DbContextOptions<MyDbContext> options)
            : base(options)
        {
        }

// This is required, because the compiler can't statically prove the EF populates
// these in the constructor.
#nullable disable
        /// <summary>
        /// Gets or sets the <see cref="MyDataModel" /> entities.
        /// </summary>
        public DbSet<MyDataModel> MySet { get; set; }
#nullable restore

        /// <summary>
        /// Configure data models.
        /// </summary>
        /// <param name="modelBuilder">The builder being used to construct the model for this context.</param>
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<MyDataModel>(b =>
            {
                b.HasKey(dm => dm.Id);
                b.Property(dm => dm.Data)
                    .IsRequired();
            });
        }
    }
}
```

### `MyDataModel.cs`

```C#
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;

namespace test_efcore_inmemory
{
    [Table("table_name", Schema = "schema_name")]
    public class MyDataModel
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="MyDataModel"/> class.
        /// </summary>
        /// <param name="id">id</param>
        /// <param name="data">data</param>
        public MyDataModel(Guid id, int data)
        {
            Id = id;
            Data = data;
        }

        public Guid Id { get; }

        public int Data { get; }
    }
}
```

### `Program.cs`

```C#
using Microsoft.EntityFrameworkCore;
using System;
using System.Threading.Tasks;

namespace test_efcore_inmemory
{
    class Program
    {
        static void Main(string[] args)
        {
            DoIt().Wait();
        }

        public static async Task DoIt()
        {
            // Arrange
            var options = new DbContextOptionsBuilder<MyDbContext>()
                .UseInMemoryDatabase("my_database")
                .Options;
            var id = Guid.NewGuid();
            var dm1 = new MyDataModel(id, 1);
            var dm2 = new MyDataModel(id, 2);

            using (var context = new MyDbContext(options))
            {
                System.Console.WriteLine("Saved {0}, {1}", dm1.Id, dm1.Data);
                context.Add(dm1);
                context.SaveChanges();
            }

            using (var context = new MyDbContext(options))
            {
                // workaround bug
                // await context.MySet.FirstAsync();

                try
                {
                    var entry = context.Entry<MyDataModel>(dm2);
                    if (entry.State.Equals(EntityState.Unchanged))
                    {
                        throw new InvalidOperationException();
                    }

                    context.Add(dm2);
                    await context.SaveChangesAsync().ConfigureAwait(false);
                }
                catch (Exception)
                {
                    System.Console.WriteLine("Failed to save {0}, {1}", dm2.Id, dm2.Data);
                    System.Console.WriteLine("Error: {0}", e);
                }

                var foundDm = await context.MySet.FirstAsync(dm => dm.Id == id);
                System.Console.WriteLine("Found {0}, {1}", foundDm.Id, foundDm.Data);
            }
        }
    }
}
```

## Output

With the workaround commented out, the output is:

```
Saved f227bcbe-231f-4ae1-a4cd-5416672fd892, 1
Failed to save f227bcbe-231f-4ae1-a4cd-5416672fd892, 2
Error: System.ArgumentException: An item with the same key has already been added. Key: f227bcbe-231f-4ae1-a4cd-5416672fd892
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryTable`1.Create(IUpdateEntry entry)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryStore.ExecuteTransaction(IList`1 entries, IDiagnosticsLogger`1 updateLogger)
   at Microsoft.EntityFrameworkCore.InMemory.Storage.Internal.InMemoryDatabase.SaveChangesAsync(IList`1 entries, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IList`1 entriesToSave, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(DbContext _, Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)
   at test_efcore_inmemory.Program.DoIt() in /tmp/test_efcore_inmemory/Program.cs:line 45
Found f227bcbe-231f-4ae1-a4cd-5416672fd892, 2
```

With the workaround not commented out, the output is:

```
Saved f631e042-332d-45bf-9777-baa13ad1c64f, 1
Failed to save f631e042-332d-45bf-9777-baa13ad1c64f, 2
Error: System.InvalidOperationException: The instance of entity type 'MyDataModel' cannot be tracked because another instance with the same key value for {'Id'} is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached. Consider using 'DbContextOptionsBuilder.EnableSensitiveDataLogging' to see the conflicting key values.
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.ThrowIdentityConflict(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.Add(TKey key, InternalEntityEntry entry, Boolean updateDuplicate)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.Add(TKey key, InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.Add(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.StartTracking(InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges, Boolean modifyProperties)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges, Boolean modifyProperties, Nullable`1 forceStateWhenUnknownKey)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityGraphAttacher.PaintAction(EntityEntryGraphNode`1 node)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityEntryGraphIterator.TraverseGraph[TState](EntityEntryGraphNode`1 node, Func`2 handleNode)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.EntityGraphAttacher.AttachGraph(InternalEntityEntry rootEntry, EntityState targetState, EntityState storeGeneratedWithKeySetTargetState, Boolean forceStateWhenUnknownKey)
   at Microsoft.EntityFrameworkCore.DbContext.SetEntityState(InternalEntityEntry entry, EntityState entityState)
   at Microsoft.EntityFrameworkCore.DbContext.SetEntityState[TEntity](TEntity entity, EntityState entityState)
   at Microsoft.EntityFrameworkCore.DbContext.Add[TEntity](TEntity entity)
   at test_efcore_inmemory.Program.DoIt() in /tmp/test_efcore_inmemory/Program.cs:line 44
Found f631e042-332d-45bf-9777-baa13ad1c64f, 1
```

## Provider and version information

EF Core version: 5.0.3
Database provider: Microsoft.EntityFrameworkCore.InMemory
Target framework: .NET 5.0
Operating system: linux

**Comment by ajcvickers** [comment]
TL;DR This is a duplicate of https://github.com/dotnet/EntityFramework.Docs/issues/709

Here's what's happening. The difference in behavior depends on whether or not the context is already tracking an entity with the given key. If the context is not already tracking an entity with the given key, then calling `Add` results in the given entity being tracked as Added:

```
MyDataModel {Id: e177f7cd-3449-43ee-ae6b-26af3764d572} Added
  Id: 'e177f7cd-3449-43ee-ae6b-26af3764d572' PK
  Data: 2
```

The in-memory database throws when attempting to save this entity because there is already an entity saved with this key value. The context at this point is still tracking the entity in the same state:

```
MyDataModel {Id: e177f7cd-3449-43ee-ae6b-26af3764d572} Added
  Id: 'e177f7cd-3449-43ee-ae6b-26af3764d572' PK
  Data: 2
```

Now attempting to query for an entity with that key on the _same context instance_ will not bring the entity back from the database, since EF Core is already tracking an entity with this key. Therefore, [identity resolution](https://docs.microsoft.com/en-us/ef/core/change-tracking/identity-resolution) kicks in and the tracked entity is returned.

If you spin up a new context and query again you will see that the in-memory database has not been updated:

```C#
using (var context = new MyDbContext(options))
{
    var foundDm = await context.MySet.FirstAsync(dm => dm.Id == id);
    System.Console.WriteLine("Found {0}, {1}", foundDm.Id, foundDm.Data);
}
```
Output:
```
Found 2e22049c-b655-4e39-bb83-b606e9b3de1f, 1
```

Now, switching to the case where the context is already tracking an entity with the given key value. In this case, Add fails immediately because EF cannot track two instances with the same key. Since this is an `InvalidOperationException` it indicates a program error. That is, it is not valid for an application to attempt to track two instances with the same key. This isn't something that can generally be recovered from, hence https://github.com/dotnet/EntityFramework.Docs/issues/709

**Comment by amosonn** [comment]
Thanks for the detailed reply! I now understand what's going on, and probably uncovered a bug in our code ;)

Is it really impossible to recover from the error of `Add`-ing two instances with the same key? If so, what's the recommended approach in case the context might have been used before - always checking the local set before calling `Add`?

**Comment by ajcvickers** [comment]
> Is it really impossible to recover from the error of Add-ing two instances with the same key?

You should not rely on being able to continue to use the context.

> If so, what's the recommended approach in case the context might have been used before - always checking the local set before calling Add?

It is recommended that your business logic ensure valid key values. Also, it is recommended that you use a short-lived DbContext instance for a single unit-of-work, in which case the context will never have been used before.
